### ISSUE TITLE
dired icons not enabled before activating treemacs

### ISSUE THREAD

**Issue by memeplex** [Alexander-Miller/treemacs#860]
I've installed the dired icons extension as explained in https://github.com/Alexander-Miller/treemacs#installation:

```elisp
(use-package treemacs
  :bind ("C-c e" . treemacs))
(use-package treemacs-icons-dired
  :after (treemacs dired)
  :config (treemacs-icons-dired-mode))
```

Notice that I'm not using `:ensure t` because I use straight.el and also that `:defer t` is not necessary because of the `:bind` clause which implies it by default.

If after starting emacs I open a dired buffer the icons are not shown. If OTOH I open treemacs and then open another dired buffer or refresh the first one, the icons appear.

Forcing treemacs loading with `:demand t` or `:defer nil` fixes the issue albeit in a less than ideal way.

If instead I change the `:after` clause to only require dired the problem also vanishes and I get no "loading" log from use-package, so I assume it's still deferring loading. As soon as I do `C-x d <Return>` I see the following in `*Messages*`:

```
Loading package treemacs-icons-dired...
Loading package treemacs-all-the-icons...
Configuring package treemacs-all-the-icons...done
Loading package treemacs-all-the-icons...done
Configuring package treemacs-icons-dired...done
Loading package treemacs-icons-dired...done (0.346s)
Configuring package dired...done
```

It seems to me this is a proper fix, to just require `:after dired` so that dired forces treemacs loading, otherwise the extension won't function properly until treemacs is loaded by an explicit user action.

**Comment by Alexander-Miller** [comment]
You're right, dired-icons should be set to load after just dired, and this is a bug in the documentation.

**Comment by memeplex** [comment]
Thanks!

**Comment by memeplex** [comment]
I've been playing a bit with this because I realized that many parts of emacs trigger the loading of dired, even if no dired buffer is actually shown, so nothing is really deferred or perhaps it's deferred until you make some unlucky change to your init.el. For what it's worth the following method is more robust in order to ensure deferral of loading:

```elisp
(use-package treemacs-icons-dired
  :init
  (defun my-treemacs-icons-dired-mode-hook ()
    (treemacs-icons-dired-mode)
    (remove-hook 'dired-mode-hook #'my-treemacs-icons-dired-mode-hook))
  :hook (dired-mode . my-treemacs-icons-dired-mode-hook))
```

Since treemacs-icons-dired-mode is a global mode I took care of removing the hook after its first execution.

I don't know why this matters to me since I run emacs in client/server mode, but it's easy to get obsessed with startup times, specially when you have use-package verbose log to look at.
